File: docs/examples/public_edge_client.py
#!/usr/bin/env python3

import argparse
import json
import time
import urllib.error
import urllib.parse
import urllib.request


def http_json(url, api_key=None, method="GET", body=None, timeout=3.0):
    headers = {"Accept": "application/json"}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    if body is not None:
        headers["Content-Type"] = "application/json"
        body = json.dumps(body).encode("utf-8")
    request = urllib.request.Request(url, method=method, headers=headers, data=body)
    started = time.perf_counter()
    with urllib.request.urlopen(request, timeout=timeout) as response:
        latency_ms = (time.perf_counter() - started) * 1000.0
        return json.load(response), latency_ms


def normalize_base_url(url):
    return url.rstrip("/")


def discover_edges(bootstraps):
    seen = {}
    for base_url in bootstraps:
        base_url = normalize_base_url(base_url)
        seen[base_url] = {"base_url": base_url}
        try:
            payload, _ = http_json(f"{base_url}/gateway/edges")
        except (urllib.error.URLError, urllib.error.HTTPError, TimeoutError, ValueError):
            continue
        for edge in payload.get("edges", []):
            edge_base_url = normalize_base_url(edge.get("base_url", ""))
            if edge_base_url:
                seen[edge_base_url] = edge
    return list(seen.values())


def probe_edge(edge):
    base_url = normalize_base_url(edge["base_url"])
    try:
        health, latency_ms = http_json(f"{base_url}/gateway/health")
    except (urllib.error.URLError, urllib.error.HTTPError, TimeoutError, ValueError):
        return None
    return {
        "edge_id": health.get("edge_id") or edge.get("edge_id") or base_url,
        "region": health.get("region") or edge.get("region") or "unknown",
        "base_url": base_url,
        "latency_ms": latency_ms,
        "ready": bool(health.get("ready")),
        "overloaded": bool(health.get("overloaded", False)),
        "healthy_nodes": int(health.get("healthy_nodes", 0)),
    }


def edge_sort_key(candidate):
    return (
        not candidate["ready"],
        candidate["overloaded"],
        -candidate["healthy_nodes"],
        candidate["latency_ms"],
    )


def choose_edge(bootstraps):
    discovered = discover_edges(bootstraps)
    candidates = [probe_edge(edge) for edge in discovered]
    candidates = [candidate for candidate in candidates if candidate is not None]
    if not candidates:
        raise RuntimeError("no reachable public edges")
    candidates.sort(key=edge_sort_key)
    return candidates[0], candidates


def main():
    parser = argparse.ArgumentParser(description="Pick the best shardd public edge and send a request.")
    parser.add_argument("--bootstrap", action="append", required=True, help="Bootstrap edge base URL.")
    parser.add_argument("--api-key", required=True, help="Bearer API key.")
    parser.add_argument("--bucket", required=True)
    parser.add_argument("--account", default="main")
    parser.add_argument("--amount", type=int, default=10)
    parser.add_argument("--write", action="store_true", help="POST /events instead of GET /balances.")
    args = parser.parse_args()

    selected, candidates = choose_edge(args.bootstrap)
    print(json.dumps({"selected_edge": selected, "candidates": candidates}, indent=2))

    if args.write:
        payload = {
            "bucket": args.bucket,
            "account": args.account,
            "amount": args.amount,
            "note": "python example",
        }
        response, _ = http_json(
            f"{selected['base_url']}/events",
            api_key=args.api_key,
            method="POST",
            body=payload,
            timeout=5.0,
        )
    else:
        query = urllib.parse.urlencode({"bucket": args.bucket})
        response, _ = http_json(
            f"{selected['base_url']}/balances?{query}",
            api_key=args.api_key,
            timeout=5.0,
        )
    print(json.dumps(response, indent=2))


if __name__ == "__main__":
    main()

File: docs/examples/public_edge_client.ts
declare const process: {
  env: Record<string, string | undefined>;
  argv: string[];
  exit(code?: number): never;
};

type EdgeDirectoryResponse = {
  observed_at_unix_ms: number;
  edges: PublicEdgeSummary[];
};

type PublicEdgeSummary = {
  edge_id: string;
  region: string;
  base_url: string;
  health_url: string;
  reachable: boolean;
  ready: boolean;
  observed_at_unix_ms?: number;
  discovered_nodes?: number;
  healthy_nodes?: number;
  best_node_rtt_ms?: number;
  sync_gap?: number;
  overloaded?: boolean;
};

type PublicEdgeHealthResponse = {
  observed_at_unix_ms: number;
  edge_id?: string;
  region?: string;
  base_url?: string;
  ready: boolean;
  discovered_nodes: number;
  healthy_nodes: number;
  best_node_rtt_ms?: number;
  sync_gap?: number;
  overloaded?: boolean;
  auth_enabled: boolean;
};

type Candidate = {
  edgeId: string;
  region: string;
  baseUrl: string;
  latencyMs: number;
  ready: boolean;
  overloaded: boolean;
  healthyNodes: number;
};

function normalizeBaseUrl(url: string): string {
  return url.replace(/\/+$/, "");
}

async function fetchJson<T>(
  url: string,
  init?: RequestInit,
): Promise<{ payload: T; latencyMs: number }> {
  const started = Date.now();
  const response = await fetch(url, init);
  const latencyMs = Date.now() - started;
  if (!response.ok) {
    throw new Error(`${url} -> ${response.status}`);
  }
  return {
    payload: (await response.json()) as T,
    latencyMs,
  };
}

async function discoverEdges(bootstraps: string[]): Promise<PublicEdgeSummary[]> {
  const seen = new Map<string, PublicEdgeSummary>();
  for (const rawBaseUrl of bootstraps) {
    const baseUrl = normalizeBaseUrl(rawBaseUrl);
    seen.set(baseUrl, {
      edge_id: baseUrl,
      region: "unknown",
      base_url: baseUrl,
      health_url: `${baseUrl}/gateway/health`,
      reachable: false,
      ready: false,
    });
    try {
      const { payload } = await fetchJson<EdgeDirectoryResponse>(`${baseUrl}/gateway/edges`);
      for (const edge of payload.edges) {
        seen.set(normalizeBaseUrl(edge.base_url), edge);
      }
    } catch {
      continue;
    }
  }
  return [...seen.values()];
}

async function probeEdge(edge: PublicEdgeSummary): Promise<Candidate | null> {
  const baseUrl = normalizeBaseUrl(edge.base_url);
  try {
    const { payload, latencyMs } = await fetchJson<PublicEdgeHealthResponse>(
      `${baseUrl}/gateway/health`,
    );
    return {
      edgeId: payload.edge_id ?? edge.edge_id ?? baseUrl,
      region: payload.region ?? edge.region ?? "unknown",
      baseUrl,
      latencyMs,
      ready: payload.ready,
      overloaded: Boolean(payload.overloaded),
      healthyNodes: payload.healthy_nodes,
    };
  } catch {
    return null;
  }
}

function compareCandidates(a: Candidate, b: Candidate): number {
  if (a.ready !== b.ready) {
    return a.ready ? -1 : 1;
  }
  if (a.overloaded !== b.overloaded) {
    return a.overloaded ? 1 : -1;
  }
  if (a.healthyNodes !== b.healthyNodes) {
    return b.healthyNodes - a.healthyNodes;
  }
  return a.latencyMs - b.latencyMs;
}

async function chooseBestEdge(bootstraps: string[]): Promise<{ selected: Candidate; candidates: Candidate[] }> {
  const edges = await discoverEdges(bootstraps);
  const candidates = (await Promise.all(edges.map((edge) => probeEdge(edge)))).filter(
    (candidate): candidate is Candidate => candidate !== null,
  );
  if (candidates.length === 0) {
    throw new Error("no reachable public edges");
  }
  candidates.sort(compareCandidates);
  return { selected: candidates[0], candidates };
}

async function main(): Promise<void> {
  const apiKey = process.env.SHARDD_API_KEY;
  if (!apiKey) {
    throw new Error("set SHARDD_API_KEY");
  }

  const bootstraps = process.argv.slice(2);
  if (bootstraps.length === 0) {
    throw new Error("usage: tsx public_edge_client.ts <bootstrap-url> [more-bootstrap-urls]");
  }

  const bucket = process.env.SHARDD_BUCKET ?? "orders";

  const { selected, candidates } = await chooseBestEdge(bootstraps);
  console.log(JSON.stringify({ selected, candidates }, null, 2));

  const balancesUrl = new URL(`${selected.baseUrl}/balances`);
  balancesUrl.searchParams.set("bucket", bucket);

  const response = await fetch(balancesUrl, {
    headers: {
      Authorization: `Bearer ${apiKey}`,
      Accept: "application/json",
    },
  });
  if (!response.ok) {
    throw new Error(`balances request failed: ${response.status}`);
  }

  console.log(JSON.stringify(await response.json(), null, 2));
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
